### PR TITLE
Fix an issue of cluster mode transition not terminating k3s

### PR DESCRIPTION
**Pull request by naiming-zededa**
# Description

Fix an issue seen recently the K3S was not terminated during transition

## How to test and validate this PR

- need to compile with 'kubevirt' tag, and bring up the device. then transition to cluuster mode. Make sure at least on the 'seed server', the transition to cluster mode is successful. Meaning the 'kubectl show node' will display the 'etcd' in the ROLES.
- although such an issue of not terminating the K3S process is a corner case, and it rarely happens to hit during the transition.

Test Done:
I have tested this patch in my own kubevirt setup, does not normal cluster transition bring up. It works fine.

## Changelog notes

Fix an issue of cluster mode transition not terminating k3s process

## Checklist

- [x] I've provided a proper description
- [ ] I've added the proper documentation
- [x] I've tested my PR on amd64 device
- [ ] I've tested my PR on arm64 device
- [ ] I've written the test verification instructions
- [ ] I've set the proper labels to this PR

### CODE REVIEW

**Review by OhmSpectator — CHANGES_REQUESTED** [review]
The PR description still contains template placeholders, which makes it hard to review and understand the intent clearly.

I’d really appreciate it if we could take a bit more care with the PR descriptions, it’s not just a formality. A well-written description shows respect for the reviewer’s time and helps us all maintain the quality of the project. Please consider updating it accordingly.

**Comment by naiming-zededa** [comment]
description updated.

**Comment by rene** [comment]
@naiming-zededa , as @OhmSpectator pointed out, it's really important to have the PR description filled according to the template, specially because it helps to run scripts to fetch information and help maintainers create the documentation for testing and validation. I still see some issues with this PR:

image

Major:
* Mixed text of the commits before the description
* No changelog information was provided (I think it's needed in this case since you are not fixing CI/CD stuff, build stuff, etc, that wouldn't require changelog notes)

Minor:
* Empty sections, you can remove them
* You don't need to put spaces in the items `[x]` so it gets correctly rendering by the GH page.

I'm fixed everything, but you still need to provide the Changelog information, please.

File: pkg/kube/cluster-utils.sh [hunk 1]
```diff
@@ -224,25 +227,62 @@ get_eve_os_release() {
...
+            kill "$pid" 2>/dev/null
+        else
+            # Subsequent attempts: use SIGKILL for forced termination
+            logmsg "Sending SIGKILL to PID $pid"
+            kill -9 "$pid" 2>/dev/null
+        fi
+    done
+
+    # Wait a moment for processes to terminate
+    sleep 1
```
Comment (rene) [inline pkg/kube/cluster-utils.sh:261]: There are several sleeps on this script, each one waiting random times, several waits 5 seconds, some 2, some 1.... here we have 1sec to wait processes to finish, is this enough?

File: pkg/kube/cluster-init.sh [hunk 2]
```diff
@@ -204,6 +205,15 @@ config_cluster_roles() {
 }
 
 check_start_k3s() {
+  # If cluster is in transition, wait until transition is complete
+  if [ "$WAIT_FOR_CLUSTER_TRANSITION" = "true" ]; then
+        logmsg "Cluster transition in progress, waiting before starting k3s"
+        while [ "$WAIT_FOR_CLUSTER_TRANSITION" = "true" ]; do
+                sleep 2
```
Comment (rene) [inline pkg/kube/cluster-init.sh]: FWIW, you can use named pipes to implement this kind of notification, instead of sleep over variables... for example:

```sh
#!/bin/sh

NP="/tmp/fifo$$"
mkfifo "$NP"

funcA() {
	echo "Waiting for funcB()..."
	read lock < "$NP"
	echo "funcA: Done"
}

funcB() {
	echo "Sleeping before notify funcA..."
	sleep 5
	echo "WAKEUP!" > "$NP"
	echo  "funcB: Done"
}

funcB &
funcA

rm "$NP"
```

File: pkg/kube/cluster-utils.sh [hunk 3]
```diff
@@ -224,25 +227,62 @@ get_eve_os_release() {
...
-    logmsg "'k3s server' process not found"
+    for pid in $pids; do
+        if [ $attempt -eq 0 ]; then
+            # First attempt: use SIGTERM for graceful shutdown
+            logmsg "Sending SIGTERM to PID $pid for graceful shutdown"
+            kill "$pid" 2>/dev/null
+        else
+            # Subsequent attempts: use SIGKILL for forced termination
+            logmsg "Sending SIGKILL to PID $pid"
+            kill -9 "$pid" 2>/dev/null
```
Comment (rene) [inline pkg/kube/cluster-utils.sh:256]: You just try SIGTERM once in the first attempt, then there is no reason to send SIGKILL signal multiple times, so no need to try 5 attempts... I suggest that you try to send SIGTERM at least 3 times, and if doesn't work, then you send a single SIGKILL and that's enough... i.e., maximum of 4 attempts...

File: pkg/kube/cluster-utils.sh [hunk 4]
```diff
@@ -224,25 +227,62 @@ get_eve_os_release() {
...
+        fi
+    done
+
+    # Wait a moment for processes to terminate
+    sleep 1
+    attempt=$((attempt+1))
+  done
+
+  # Final check for any remaining processes
+  remaining_pids=$(pgrep -f "$K3S_SERVER_CMD")
```
Comment (rene) [inline pkg/kube/cluster-utils.sh]: I don't understand why you need another check here... if processes are not getting killed immediately with SIGKILL then most probably they are blocked on some I/O operation (which might be not only storage, but network as well, for instance), in the code above you try to send at maximum of 4x times a SIGKILL with a 1sec sleep, and then, sleeps for 1 second again and try to send another one... no difference if you just increase the attempts to 6, instead of 5... but see my last comment...

**Comment by naiming-zededa** [comment]
> 
> I just fixed everything, but you still need to provide the Changelog information, please.

Got it, thanks @rene

File: pkg/kube/cluster-utils.sh (hunk 1 above)
Comment (naiming-zededa) [reply pkg/kube/cluster-utils.sh:261]: I think in this loop, we want to check if the kill taking effect, one second here should be good.
The problem turned out in the case we found, actually was not the kill itself didn't kill, but we didn't know there was a case the pid can be more than one. We passed in a wrong string to kill. Since we are fixing the multiple pids case, we adding the loop to be sure with '-9'.

File: pkg/kube/cluster-init.sh (hunk 2 above)
Comment (naiming-zededa) [reply pkg/kube/cluster-init.sh]: good suggestion, will do.

File: pkg/kube/cluster-utils.sh (hunk 3 above)
Comment (naiming-zededa) [reply pkg/kube/cluster-utils.sh:256]: Sure.

File: pkg/kube/cluster-utils.sh (hunk 4 above)
Comment (naiming-zededa) [reply pkg/kube/cluster-utils.sh]: ok removed.

**Comment by naiming-zededa** [comment]
update @rene, please review.

**Comment by naiming-zededa** [comment]
> @naiming-zededa try rebasing on master to see if that fixes the kubevirt build issue

rebased to see.

**Comment by rene** [comment]
@naiming-zededa , LGTM but you need to rebase...

**Comment by naiming-zededa** [comment]
@rene rebased. thanks.
